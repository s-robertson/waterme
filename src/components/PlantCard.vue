<template>
  <div class="plant-card">
    <div class="plant-card__options">
      <label class="plant-card__label">
        <input :checked="isChecked" type="checkbox" @click="handleChange" />
        {{ plant.name }}
      </label>
      <router-link :to="`edit-plant/${plant.id}`">
        <SettingsIcon
          class="plant-card__settings"
          :alt="`Edit ${plant.name}`"
        />
      </router-link>
    </div>
    <div class="plant-card__watering-progress">
      <div class="plant-card__watering-progress-text">
        {{ `${daysRemaining} Days Remaining` }}
      </div>
      <div class="plant-card__watering-progress-fill" :style="progressStyle" />
    </div>
  </div>
</template>

<script>
import moment from "moment";
import SettingsIcon from "@/assets/settings.svg";

export default {
  name: "PlantCard",
  props: {
    plant: Object,
    isChecked: Boolean
  },
  computed: {
    daysRemaining() {
      const daysBetweenWatering = this.plant.days;
      const lastWatered = moment
        .unix(this.plant.lastWatered)
        .set({ second: 0, minute: 0, hour: 0 });
      const now = moment().set({ second: 0, minute: 0, hour: 0 });
      const daysSinceLastWatered = now.diff(lastWatered, "days");
      const daysRemaining = daysBetweenWatering - daysSinceLastWatered;

      if (daysRemaining > daysBetweenWatering) {
        return daysBetweenWatering;
      }

      return daysRemaining > 0 ? daysRemaining : 0;
    },
    wateringProgress() {
      return (this.daysRemaining / this.plant.days) * 100;
    },
    progressStyle() {
      return {
        width: `${this.wateringProgress}%`
      };
    }
  },
  methods: {
    handleChange(e) {
      this.$emit("toggle", e.target.checked);
    }
  },
  components: {
    SettingsIcon
  }
};
</script>

<style lang="scss">
@import "../assets/variables";

.plant-card__options {
  margin-bottom: 1rem;
}

.plant-card__settings {
  width: 25px;
  cursor: pointer;
  color: #00dcff;
  fill: $link-primary;
  float: right;
  vertical-align: middle;
}

.plant-card__watering-progress {
  border: 2px solid #fff;
  position: relative;
  height: 2rem;
  border-radius: 10px;
}

.plant-card__watering-progress-fill {
  background: $brand-secondary;
  height: 2rem;
  border-radius: 10px;
}

.plant-card__watering-progress-text {
  position: absolute;
  width: 100%;
  text-align: center;
  height: 1rem;
  padding: 0.5rem 0;
  box-sizing: content-box;
}
</style>
