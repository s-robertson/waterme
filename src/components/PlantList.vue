<template>
  <div class="plant-list">
    <p v-if="!hasPlants">
      You haven't added any plants yet!
    </p>
    <div v-if="hasPlants">
      <p>Select plants to water:</p>
      <form action="" @submit.prevent="handleSubmit">
        <PlantCard
          v-for="(plant, plantIndex) in sortedPlants"
          :key="plantIndex"
          :is-checked="isChecked(plantIndex)"
          @toggle="toggled => handlePlantToggle(toggled, plantIndex)"
          :plant="plant"
          class="plant-list__plant"
        />
        <button type="submit">Water Selected Plants</button>
        <router-link class="button" to="/add-plant"
          >Add a new plant</router-link
        >
      </form>
    </div>
  </div>
</template>

<script>
import { mapActions, mapGetters } from "vuex";
import PlantCard from "@/components/PlantCard";

export default {
  name: "PlantList",
  components: { PlantCard },
  data() {
    return {
      plantsToWater: []
    };
  },
  computed: {
    ...mapGetters(["sortedPlants"]),
    hasPlants() {
      return this.sortedPlants.length > 0;
    }
  },
  methods: {
    ...mapActions(["waterPlants"]),
    isChecked(plantId) {
      return this.plantsToWater.findIndex(item => item === plantId) !== -1;
    },
    handleSubmit() {
      if (this.plantsToWater.length) {
        this.waterPlants(this.plantsToWater);
        this.plantsToWater = [];
      }
    },
    handlePlantToggle(toggled, id) {
      if (toggled) {
        this.plantsToWater = [...this.plantsToWater, id];
        return;
      }
      this.plantsToWater = this.plantsToWater.filter(plantId => plantId !== id);
    }
  }
};
</script>

<style>
.plant-list__plant {
  margin-bottom: 3rem;
}
</style>
