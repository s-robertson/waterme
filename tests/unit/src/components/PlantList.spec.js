import { shallowMount, createLocalVue } from "@vue/test-utils";
import PlantList from "@/components/PlantList";
import Vuex from "vuex";
import PlantCard from "@/components/PlantCard";

const localVue = createLocalVue();

localVue.use(Vuex);

describe("PlantForm component", () => {
  let mockPlants;

  beforeEach(() => {
    mockPlants = [
      {
        name: "Plant One",
        days: 10
      },
      {
        name: "Plant Two",
        days: 10
      },
      {
        name: "Plant Three",
        days: 10
      }
    ];
  });

  it("Should render correctly when plants have not been added", () => {
    const store = new Vuex.Store({
      state: {},
      getters: {
        sortedPlants() {
          return [];
        }
      }
    });

    const wrapper = shallowMount(PlantList, {
      store,
      localVue,
      stubs: ["router-link"]
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("Should render correctly when plants have been added", () => {
    const store = new Vuex.Store({
      state: {},
      getters: {
        sortedPlants() {
          return mockPlants;
        }
      }
    });
    const wrapper = shallowMount(PlantList, {
      store,
      localVue,
      stubs: ["router-link"]
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("Should keep track of which plants to water", () => {
    const store = new Vuex.Store({
      state: {},
      getters: {
        sortedPlants() {
          return mockPlants;
        }
      }
    });
    const wrapper = shallowMount(PlantList, {
      store,
      localVue,
      stubs: ["router-link"]
    });

    const cardComponents = wrapper.findAllComponents(PlantCard);
    cardComponents.at(0).vm.$emit("toggle", true);
    cardComponents.at(1).vm.$emit("toggle", true);
    cardComponents.at(1).vm.$emit("toggle", false);
    cardComponents.at(2).vm.$emit("toggle", true);
    expect(wrapper.vm.plantsToWater).toEqual([0, 2]);
  });

  it("Should mark selected plants as watered", () => {
    const mockActions = {
      waterPlants: jest.fn()
    };

    const store = new Vuex.Store({
      state: {},
      actions: mockActions,
      getters: {
        sortedPlants() {
          return mockPlants;
        }
      }
    });
    const wrapper = shallowMount(PlantList, {
      store,
      localVue,
      stubs: ["router-link"]
    });

    const cardComponents = wrapper.findAllComponents(PlantCard);
    cardComponents.at(0).vm.$emit("toggle", true);
    cardComponents.at(2).vm.$emit("toggle", true);
    wrapper.find("form").trigger("submit");
    expect(mockActions.waterPlants).toHaveBeenCalledWith(expect.any(Object), [
      0,
      2
    ]);
  });
});
