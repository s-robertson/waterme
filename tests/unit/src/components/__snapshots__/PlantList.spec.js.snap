// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlantForm component Should render correctly when plants have been added 1`] = `
<div class="plant-list">
  <!---->
  <div>
    <p>Select plants to water:</p>
    <form action="">
      <plantcard-stub plant="[object Object]" class="plant-list__plant"></plantcard-stub>
      <plantcard-stub plant="[object Object]" class="plant-list__plant"></plantcard-stub>
      <plantcard-stub plant="[object Object]" class="plant-list__plant"></plantcard-stub> <button type="submit" class="plant-list__water-button">
        Water Selected Plants
      </button>
    </form>
  </div>
  <router-link-stub to="/add-plant" class="plant-list__add-plant button">Add a new plant</router-link-stub>
</div>
`;

exports[`PlantForm component Should render correctly when plants have not been added 1`] = `
<div class="plant-list">
  <p>
    You haven't added any plants yet!
  </p>
  <!---->
  <router-link-stub to="/add-plant" class="plant-list__add-plant button">Add a new plant</router-link-stub>
</div>
`;
